packages/contractkit/src/wallets/signers/aws-hsm-signer.ts

```diff
@@ -0,0 +1,109 @@
+import { ensureLeading0x, publicKeyToAddress, trimLeading0x } from '@celo/utils/lib/address'
+import { KMS } from 'aws-sdk'
+import BigNumber from 'bignumber.js'
+import { ec as EC } from 'elliptic'
+import * as ethUtil from 'ethereumjs-util'
+import { parseBERSignature } from '../../utils/ber-utils'
+import { bigNumberToBuffer, bufferToBigNumber, isCanonical } from '../../utils/signature-utils'
+import { getHashFromEncoded, RLPEncodedTx } from '../../utils/signing-utils'
+import { Signer } from './signer'
+
+const SigningAlgorithm = 'ECDSA_SHA_256'
+const secp256k1Curve = new EC('secp256k1')
+
+function getRecoveryParam(message: Buffer, expectedAddress: string, r: Buffer, s: Buffer): number {
+  for (let i = 0; i < 2; i++) {
+    const recoveredPublicKeyByteArray = ethUtil.ecrecover(message, i + 27, r, s)
+    const publicKeyBuff = Buffer.from(recoveredPublicKeyByteArray)
+
+    const pub = bigNumberToBuffer(bufferToBigNumber(publicKeyBuff), 64)
+    if (publicKeyToAddress(ensureLeading0x(pub.toString('hex'))) === expectedAddress) {
+      return i
+    }
+  }
+
+  throw new Error('unable to recover public key from signature')
+}
+
+export default class AwsHsmSigner implements Signer {
+  private kms: KMS
+  private keyId: string
+  private address: string
+
+  constructor(kms: KMS, keyId: string, address: string) {
+    this.kms = kms
+    this.keyId = keyId
+    this.address = address
+  }
+
+  private async findCanonicalSignature(buffer: Buffer): Promise<{ S: BigNumber; R: BigNumber }> {
+    let S: BigNumber
+    let R: BigNumber
+
+    let flag = true
+    while (flag) {
```

Comment:
Is this signing operation non-deterministic? What causes it to become canonical?

Reply:
I believe signing is non-deterministic yes.